crawl-ref/source/mon-place.cc

```diff
@@ -794,15 +610,19 @@ static bool _valid_monster_generation_location(const mgen_data &mg,
         return false;
     }
 
-    // Check player proximity to avoid band members being placed
-    // close to the player erroneously.
-    // XXX: This is a little redundant with proximity checks in
-    // place_monster.
-    if (mg.proximity == PROX_AWAY_FROM_PLAYER
-        && grid_distance(you.pos(), mg_pos) <= LOS_RADIUS)
+    bool close_to_player = grid_distance(you.pos(), mg_pos) <= LOS_RADIUS;
+    if (mg.proximity == PROX_AWAY_FROM_PLAYER && close_to_player
+        || mg.proximity == PROX_CLOSE_TO_PLAYER && !close_to_player)
     {
         return false;
     }
+    // Check that the location is not proximal to level stairs.
+    else if (mg.proximity == PROX_AWAY_FROM_STAIRS)
+    {
+        for (distance_iterator di(mg_pos, false, LOS_RADIUS); di; ++di)
+            if (feat_is_stone_stair(grd(*di)))
```

Comment:
This applies both to up stairs and down stairs; is that intended? I would have guessed this was just about up stairs, for when a player first enters the level, and that it doesn't really matter if monsters spawn near a down stair.

Also, you probably want to consider branch entrances and/or exits.

Reply:
This prox type was added to handle level generation monsters that are generated awake so that the player wouldn't encounter these in LOS when first entering a level. So for that purpose, they should probably only affect stone stairs. It's not possible to first enter a level from a branch or a portal outside of wizard mode. Adding more restrictions increases the chance that a monster doesn't place properly (currently gets 45 tries).